libminifi/test/libtest/unit/SingleProcessorTestController.h

```diff
@@ -52,25 +61,17 @@ class SingleProcessorTestController : public TestController {
 
   core::Relationship addDynamicRelationship(std::string name);
 
- private:
-  std::shared_ptr<core::FlowFile> createFlowFile(const std::string_view content, std::unordered_map<std::string, std::string> attributes);
+  template<typename T = core::Processor>
+  T* getProcessor() const { return static_cast<T*>(processor_); }
```

Comment:
With the change of adding the unique_ptr of the root process group in the test plan and that root process group owning the processors and the controllers, retrieving the pointers to the processors became problematic in some cases. For example in test suite initialization:
```
  PushGrafanaLokiGrpcTestFixture()
      : mock_loki_("10991"),
        test_controller_(std::make_unique<PushGrafanaLokiGrpc>("PushGrafanaLokiGrpc")),
        push_grafana_loki_grpc_(test_controller_.getProcessor<PushGrafanaLokiGrpc>()) {
```
In this case initializing the processor pointer would be problematic without the `getProcessor` method, we would need to initialize them separately afterwards in the main constructor block.
Also I think it would make the diff larger, as the common pattern which currently looking like this:
```
test::SingleProcessorTestController test_controller{std::make_unique<InvokeHTTP>("InvokeHTTP")};
auto invokehttp = test_controller.getProcessor();
```
would probably need to be changed to something like this:
```
auto processor = std::make_unique<InvokeHTTP>("InvokeHTTP");
auto invokehttp = processor.get();
test::SingleProcessorTestController test_controller{std::move(processor)};
```

If you think it's worth reverting and changing the initializations, I'm okay with that too, I just wanted to add some context.

Reply:
I don't like the idea of complicating TestBase or TestController to mirror the god class complexity of FlowController. I would rather keep the process groups and the processor ownership outside of that, in the test case scopes, if that's feasible.

Reply:
That's true, we shouldn't have a god class like FlowController here, but I'm not sure this change complicates the previous behavior. Previously we also had all the controllers and processors as part of the TestPlan, the only difference is that the exclusive ownership is now in the test plan instead of the shared ownership between the TestPlan and the test case. The change had to be made because controller services are now only valid on a process group level and both the processors and controllers had to be added to the same process group in the test scenarios to be able to retrieve the controllers in the scope of a processor that defines a controller in their processor properties. I think this was the least intrusive change to be able to use the processors and controllers in our test framework, of course we can redesign the TestBase, TestController and TestPlan, but I think that's out of the scope of this PR.